mm2src/coins/tx_history_storage/mod.rs

```diff
@@ -0,0 +1,183 @@
+use crate::my_tx_history_v2::TxHistoryStorage;
+use crate::TransactionType;
+use derive_more::Display;
+use mm2_core::mm_ctx::MmArc;
+use mm2_err_handle::prelude::*;
+use num_traits::Zero;
+use primitives::hash::H160;
+use serde::de::Error;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::collections::HashSet;
+use std::iter::FromIterator;
+
+#[cfg(target_arch = "wasm32")] pub mod wasm;
+
+#[cfg(not(target_arch = "wasm32"))]
+pub mod sql_tx_history_storage_v2;
+
+#[cfg(any(test, target_arch = "wasm32"))]
+mod tx_history_v2_tests;
+
+/// Get `token_id` from the transaction type.
+/// Returns an empty `token_id` if the transaction is [`TransactionType::StandardTransfer`].
+#[inline]
+pub fn token_id_from_tx_type(tx_type: &TransactionType) -> String {
+    match tx_type {
+        TransactionType::TokenTransfer(token_id) => format!("{:02x}", token_id),
+        _ => String::new(),
+    }
+}
+
+#[derive(Debug, Display)]
+pub enum CreateTxHistoryStorageError {
+    Internal(String),
+}
+
+/// `TxHistoryStorageBuilder` is used to create an instance that implements the `TxHistoryStorage` trait.
+pub struct TxHistoryStorageBuilder<'a> {
+    ctx: &'a MmArc,
+}
+
+impl<'a> TxHistoryStorageBuilder<'a> {
+    #[inline]
+    pub fn new(ctx: &MmArc) -> TxHistoryStorageBuilder<'_> { TxHistoryStorageBuilder { ctx } }
+
+    #[inline]
+    pub fn build(self) -> MmResult<impl TxHistoryStorage, CreateTxHistoryStorageError> {
+        #[cfg(target_arch = "wasm32")]
+        return wasm::IndexedDbTxHistoryStorage::new(self.ctx);
+        #[cfg(not(target_arch = "wasm32"))]
+        sql_tx_history_storage_v2::SqliteTxHistoryStorage::new(self.ctx)
+    }
+}
+
+/// Whether transaction is unconfirmed or confirmed.
+/// Serializes to either `0u8` or `1u8` correspondingly.
+#[repr(u8)]
+#[derive(Clone, Copy, Debug)]
+pub enum ConfirmationStatus {
+    Unconfirmed = 0,
+    Confirmed = 1,
+}
+
+impl ConfirmationStatus {
+    #[inline]
+    pub fn from_block_height<Height: Zero>(height: Height) -> ConfirmationStatus {
+        if height.is_zero() {
+            ConfirmationStatus::Unconfirmed
+        } else {
+            ConfirmationStatus::Confirmed
+        }
+    }
+}
+
+impl Serialize for ConfirmationStatus {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        (*self as u8).serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for ConfirmationStatus {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let code = u8::deserialize(deserializer)?;
+        match code {
```

Comment:
I'd like to add `ConfirmationStatus::from_u8` but it's needed to add an error type like `struct UnexpectedErrorType`, derive `Display` or implement it manually. Or it can just return `Result<ConfirmationStatus, String>` instead, but we prefer not to use `String` as an error.
So is it really necessary? I'm not sure `ConfirmationStatus::from_u8` such method will be used

Reply:
Instead of dealing with those errors, you can add another variant to that enum something like `Unrecognized` and return it in case of any errors. But as you mentioned, if you think `from_u8` isn't really necessary since you don't use it anywhere else, you can keep it this way then.